File: app/src/androidTest/java/network/spiritscorp/InsulinCalculatorInstrumentedTest.java
package network.spiritscorp;

/*
 * Copyright (C) 2026 Tom Spirit
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import android.content.Context;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.platform.app.InstrumentationRegistry;
import kotlin.Pair;
import network.spiritscorp.data.AppDatabase;
import network.spiritscorp.data.DatabaseBackupManager;
import network.spiritscorp.data.InsulinRepository;
import network.spiritscorp.model.CalculationLog;
import network.spiritscorp.model.CarbUnit;
import network.spiritscorp.model.TimeOfDay;
import network.spiritscorp.model.UserSettings;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

/**
 * On-Device / Instrumentation Test Suite for Android execution environments (in Java).
 *<br><br>
 * Verifies Android Context binding, package identifier, database instantiation
 * on target device, and Android-level serialization / backup operations.
 */
@RunWith(AndroidJUnit4.class)
public class InsulinCalculatorInstrumentedTest {

    @Test
    public void testAppContextAndPackageIdentity() {
        Context appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        assertNotNull(appContext);
        assertEquals("network.spiritscorp.insulincalc", appContext.getPackageName());
    }

    @Test
    public void testRoomDatabaseInstanceCreationOnDevice() {
        Context appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        AppDatabase database = AppDatabase.getDatabase(appContext);
        assertNotNull(database);
        assertNotNull(database.calculationLogDao());
        assertNotNull(database.userSettingsDao());
    }

    @Test
    public void testRepositoryAndDeviceBackupWorkflow() {
        Context appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        AppDatabase database = AppDatabase.getDatabase(appContext);
        InsulinRepository repository = new InsulinRepository(database.calculationLogDao(), database.userSettingsDao());

        UserSettings settings = repository.getSettings();
        assertNotNull(settings);
        assertTrue(settings.getMorningFactor() > 0.0);

        CalculationLog sampleLog = new CalculationLog(
                0,
                System.currentTimeMillis(),
                "Test On-Device Mahlzeit",
                40.0,
                CarbUnit.GRAMS.getShortName(),
                40.0,
                3.33,
                4.0,
                TimeOfDay.NOON.getTitle(),
                1.0,
                3.33,
                null,
                null,
                null,
                null,
                3.33,
                3.5,
                "Notiz"
        );

        DatabaseBackupManager backupManager = new DatabaseBackupManager(database);
        String json = backupManager.exportToJson(settings, Collections.singletonList(sampleLog));
        assertTrue(json.contains("Test On-Device Mahlzeit"));
        assertTrue(json.contains("settings"));

        Pair<UserSettings, List<CalculationLog>> parsed = backupManager.parseJson(json);
        assertNotNull(parsed);
        assertNotNull(parsed.getSecond());
        assertEquals(1, parsed.getSecond().size());
        assertEquals("Test On-Device Mahlzeit", parsed.getSecond().getFirst().getMealTitle());
    }
}

File: app/src/androidTest/java/network/spiritscorp/InsulinCalculatorInstrumentedTest.kt
package network.spiritscorp

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import network.spiritscorp.data.AppDatabase
import network.spiritscorp.data.DatabaseBackupManager
import network.spiritscorp.data.InsulinRepository
import network.spiritscorp.model.CalculationLog
import network.spiritscorp.model.CarbUnit
import network.spiritscorp.model.GlucoseUnit
import network.spiritscorp.model.TimeOfDay
import network.spiritscorp.model.UserSettings
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith

/**
 * On-Device / Instrumentation Test Suite for Android execution environments.
 * 
 * Verifies Android Context binding, package identifier, database instantiation
 * on target device, and Android-level serialization / backup operations.
 */
@RunWith(AndroidJUnit4::class)
class InsulinCalculatorInstrumentedTest {

  @Test
  fun testAppContextAndPackageIdentity() {
    val appContext = InstrumentationRegistry.getInstrumentation().targetContext
    assertNotNull(appContext)
    assertEquals("network.spiritscorp.insulincalc", appContext.packageName)
  }

  @Test
  fun testRoomDatabaseInstanceCreationOnDevice() {
    val appContext = InstrumentationRegistry.getInstrumentation().targetContext
    val database = AppDatabase.getDatabase(appContext)
    assertNotNull(database)
    assertNotNull(database.calculationLogDao())
    assertNotNull(database.userSettingsDao())
  }

  @Test
  fun testRepositoryAndDeviceBackupWorkflow() = runBlocking {
    val appContext = InstrumentationRegistry.getInstrumentation().targetContext
    val database = AppDatabase.getDatabase(appContext)
    val repository = InsulinRepository(database.calculationLogDao(), database.userSettingsDao())

    val settings = repository.getSettings()
    assertNotNull(settings)
    assertTrue(settings.morningFactor > 0.0)

    val sampleLog = CalculationLog(
      mealTitle = "Test On-Device Mahlzeit",
      rawCarbInput = 40.0,
      carbUnit = CarbUnit.GRAMS.shortName,
      carbGrams = 40.0,
      beValue = 3.33,
      keValue = 4.0,
      timeOfDay = TimeOfDay.NOON.title,
      insulinFactor = 1.0,
      mealInsulin = 3.33,
      totalInsulin = 3.33,
      roundedInsulin = 3.5
    )

    val json = DatabaseBackupManager.exportToJson(settings, listOf(sampleLog))
    assertTrue(json.contains("Test On-Device Mahlzeit"))
    assertTrue(json.contains("settings"))

    val parsed = DatabaseBackupManager.parseJson(json)
    assertNotNull(parsed)
    assertEquals(1, parsed!!.second.size)
    assertEquals("Test On-Device Mahlzeit", parsed.second[0].mealTitle)
  }
}
